### ISSUE TITLE
[0.16.10] Potential regression pl.concat & Categoricals

### ISSUE THREAD

**Issue by AlexanderVanEck** [pola-rs/polars#7366]
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

Hello 👋  Thank you for maintaining polars, we've been using it to great success!

I've stumbled upon a possible regression in `pl.concat` concerning pl.Categorical. We have a function that partitions a dataframe by a group column and then returns a pl.concat dataframe of x groups. The error suggest that categoricals of different sources cannot be concatenated together. Should I figure how to set a global string cache to resolve this? or is there a better way?

I've confirmed this worked in `polars==0.16.9` and was introduced in `polars==0.16.10`.

### Reproducible example

```python
offset = 1
number_of_groups = 2
df = pl.DataFrame(
    {
        'col': [
            'group_1',
            'group_1',
            'group_2',
            'group_2',
            'group_2',
            'group_3',
            'group_3',
            'group_3',
            'group_3',
        ],
    }
)
df = df.with_columns(df['col'].cast(pl.Categorical))
df_groups = df.partition_by('col', maintain_order=True)
partition_df = pl.concat(df_groups[offset : offset + number_of_groups])
print(partition_df)

>>> E               pyo3_runtime.PanicException: should not fail: ComputeError(Borrowed("Cannot concat Categoricals coming from a different source. Consider setting a global StringCache."))

/home/aicompute/env/lib/python3.10/site-packages/polars/internals/functions.py:250: PanicException
```

### Expected behavior

```
shape: (7, 1)
┌─────────┐
│ col     │
│ ---     │
│ cat     │
╞═════════╡
│ group_2 │
│ group_2 │
│ group_2 │
│ group_3 │
│ group_3 │
│ group_3 │
│ group_3 │
└─────────┘
```

### Installed versions

<details>

```
---Version info---
Polars: 0.16.10
Index type: UInt32
Platform: Linux-5.15.49-linuxkit-x86_64-with-glibc2.31
Python: 3.10.10 (main, Feb  8 2023, 14:50:01) [GCC 9.4.0]
---Optional dependencies---
pyarrow: 11.0.0
pandas: 1.5.3
numpy: 1.24.2
fsspec: <not installed>
connectorx: <not installed>
xlsx2csv: <not installed>
deltalake: <not installed>
matplotlib: <not installed>
```

</details>

**Comment by ghuls** [comment]
This works (Polars 0.16.13):

```python
offset = 1
number_of_groups = 2
df = pl.DataFrame(
    {
        'col': [
            'group_1',
            'group_1',
            'group_2',
            'group_2',
            'group_2',
            'group_3',
            'group_3',
            'group_3',
            'group_3',
        ],
    }
)

with pl.StringCache():
    df = df.with_columns(df['col'].cast(pl.Categorical))
    df_groups = df.partition_by('col', maintain_order=True)
    partition_df = pl.concat(df_groups[offset : offset + number_of_groups])
    print(partition_df)
```

But it indeed seems like a regression.

**Comment by AlexanderVanEck** [comment]
This was resolved by upgrading past 0.18.5.

I don't see immediately which MR contributed to it being fixed; but the release notes are here. https://github.com/pola-rs/polars/releases/tag/py-0.18.5

Resolving issue. Thanks @ritchie46 !
